odl/discr/partition.py

```diff
@@ -0,0 +1,561 @@
+        The endpoints ``begin`` and ``end`` are determined from the grid
+        as follows::
+
+            begin = x[0] - (x[1] - x[0]) / 2
+            end = x[N-1] + (x[N-1] - x[N-2]) / 2
+        """
+        if not isinstance(grid, TensorGrid):
+            raise TypeError('{!r} is not a TensorGrid instance.'
+                            ''.format(grid))
+
+        # Begin and end: check correct shapes and values if given,
+        # calculate otherwise.
+        begin_axes = kwargs.pop('begin_axes', None)
+        end_axes = kwargs.pop('end_axes', None)
+
+        begin = self._calc_begin_end(grid, begin_axes, begin, 'begin')
+        end = self._calc_begin_end(grid, end_axes, end, 'end')
+
+        super().__init__()
+        self._bbox = IntervalProd(begin, end)
+        self._grid = grid
+
+    def _calc_begin_end(self, grid, vec_axes, vector, which):
+        """Return the interval product begin/end from ``grid``.
+
+        Parameters
+        ----------
+        grid : `TensorGrid`
+            Grid from which to infer
+        vec_axes : sequence of `int`
+            Dimensions in which to take the values from ``vector``
+        vector : array-like, shape ``(grid.ndim,)``
+            Array from which to take the values corresponding to the
+            dimensions in ``vec_axes``
+        which : {'begin', 'end'}
+            Which one of the endpoints to calculate
+
+        Returns
+        -------
+        newvec : `numpy.ndarray`
+            The new begin or end vector, depending on ``which``
+        """
+        # Check axes sanity
+        if vector is None:
+            vector = np.zeros(grid.ndim)
+            if vec_axes is None:
+                vec_axes = []
+            else:
+                raise ValueError('{which}_axes cannot be given without '
+                                 '{which} parameter.'.format(which=which))
+        elif vec_axes is None:
+            vec_axes = list(range(grid.ndim))
+
+        # Normalize negative values and check for errors
+        vec_axes = [i if i >= 0 else grid.ndim + i
+                    for i in vec_axes]
+        if any(i < 0 or i >= grid.ndim for i in vec_axes):
+            raise IndexError('{}_axes {} contains out-ouf-bounds indices.'
+                             ''.format(which, vec_axes))
+
+        if len(set(vec_axes)) != len(vec_axes):
+            raise ValueError('{}_axes {} contains duplicate indices.'
+                             ''.format(which, vec_axes))
+
+        # Set private attribute to be used in __repr__
+        # TODO: find good way to handle this without running into slight
+        # rounding errors. Best would be to determine in __repr__ if
+        # begin/end are as calculated, but it may be better to store if
+        # begin/end was given and just always add it to the __repr__ output
+        # in that case, regardless of other factors
+        if which == 'begin':
+            if vec_axes:
+                self._custom_begin = True
```

Comment:
Function says `Return` but it also sets values of `self`, perhaps change how this is done/named?